src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1450,6 +1450,146 @@ julia> unique!(df)  # modifies df
+unique combinations of `indexcols` values. If `allowduplicates=true`
+duplicates are allowed
+
+# Examples
+```jldoctest
+julia> df = DataFrame(x=1:2, y='a':'b', z=["x", "y"])
+2×3 DataFrame
+ Row │ x      y     z
+     │ Int64  Char  String
+─────┼─────────────────────
+   1 │     1  a     x
+   2 │     2  b     y
+
+julia> completecombinations(df, [:x, :y])
+4×3 DataFrame
+ Row │ x      y     z
+     │ Int64  Char  String?
+─────┼──────────────────────
+   1 │     1  a     x
+   2 │     2  a     missing
+   3 │     1  b     missing
+   4 │     2  b     y
+
+julia> completecombinations(df, [:y, :z], fill=0)
+4×3 DataFrame
+ Row │ x       y     z
+     │ Int64?  Char  String
+─────┼──────────────────────
+   1 │      1  a     x
+   2 │      0  b     x
+   3 │      0  a     y
+   4 │      2  b     y
+```
+"""
+function completecombinations(df::AbstractDataFrame, indexcols;
+                              allowduplicates::Bool=false, fill=missing)
+    _check_consistency(df)
+
+    colind = index(df)[indexcols]
+
+    if length(colind) == 0
+        throw(ArgumentError("At least one column to complete combinations " *
+                            "must be specified"))
+    end
+
+    has_duplicates = row_group_slots(ntuple(i -> df[!, colind[i]], length(colind)),
+                                     Val(false), nothing, false, nothing)[1] != nrow(df)
+    if has_duplicates && !allowduplicates
+        throw(ArgumentError("duplicate combinations of `indexcols` are not " *
+                            "allowed in input when `allowduplicates=false`"))
+    end
+
+    # Create a vector of vectors of unique values in each column
+    uniquevals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        # All levels are retained, missing is added only if present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[!, col]), missing)
+        else
+            tempcol = levels(df[!, col])
+        end
+        push!(uniquevals, tempcol)
+    end
+
+    # make sure we do not overflow in the target data frame size
+    target_rows = Int(prod(x -> big(length(x)), uniquevals))
+    if iszero(target_rows)
+        @assert iszero(nrow(df))
+        return copy(df)
+    end
+
+    # construct expanded columns
+    out_df = DataFrame()
+    inner = 1
+    @assert length(uniquevals) == length(colind)
+    for (val, cind) in zip(uniquevals, colind)
+        len = length(val)
+        target_col = Tables.allocatecolumn(eltype(df[!, cind]), len)
+        copy!(target_col, val)
+        last_inner = inner
+        inner *= len
+        outer, remv = divrem(target_rows, inner)
+        @assert iszero(remv)
+        out_df[!, _names(df)[cind]] = repeat(target_col, inner=last_inner, outer=outer)
+    end
+    @assert inner == target_rows
+
+    idx_ind = 0
+    while columnindex(df, string("source7249", idx_ind)) != 0
+        idx_ind += 1
+    end
```

Comment:
Maybe storing the column name in a variable and using it later instead of calling `string("source7249", idx_ind)` would be slightly cleaner? Something like
```suggestion
    idx_ind = 0
    while true
        idx_col = string("source7249", idx_ind)
        columnindex(df, idx_col) == 0 && break
        idx_ind += 1
    end
```

Same for `order_ind`.